compiler/rustc_trait_selection/src/error_reporting/infer/need_type_info.rs

```diff
@@ -613,6 +619,43 @@ impl<'a, 'tcx> TypeErrCtxt<'a, 'tcx> {
             }),
         }
     }
+
+    /// Detect the inference regression on crate `time` <= 0.3.35 and emit a more targeted error.
+    /// <https://github.com/rust-lang/rust/issues/127343>
+    // FIXME: we should figure out a more generic version of doing this, ideally in cargo itself.
+    fn detect_old_time_crate_version(
+        &self,
+        span: Option<Span>,
+        kind: &InferSourceKind<'_>,
+        // We will clear the non-actionable suggestion from the error to reduce noise.
+        infer_subdiags: &mut Vec<SourceKindSubdiag<'_>>,
+    ) -> Option<()> {
```

Comment:
Question: with this new detection logic, AFAICT we're not actually checking for specific `time` versions (which is impossible to check in the general case without assuming things about cargo and such). If that's the case, is it possible that the note

```
this is an inference error on crate `time` caused by a change in Rust 1.80.0; update `time` to version `>=0.3.36`
```

sounds too certain? i.e. can this also be caused by a not-caused-by-us inference error due to `time` itself (in a newer version)?

Reply:
If the coincidence of all of these conditions happening were to happen, it would mainly confuse the `time` maintainers and no-one else :)

We can also look at the path again and change the confidence of the wording depending on whether we find the version from the directory name or not (or try to find the `Cargo.toml`), but I'd rather we didn't do that (or if we start doing that, have a single place where we maintain a "get crate version from path" logic, as there *are* a bunch of places where it'd be useful to know the crate version for a given type, but I feel all of this is beyond the scope of "get something in front of users quickly").

Reply:
Yeah I'd rather not probe for `Cargo.toml` since that's not really our concern (as in rustc) I feel like 💀. I'm okay with accepting this version of detection logic if the team decides to.

Reply:
> If the coincidence of all of these conditions happening were to happen, it would mainly confuse the time maintainers and no-one else :)

For certainty, I'm fine with that :slightly_smiling_face: